enhancements/installer/aws-permissions-check-bypass.md

```diff
@@ -0,0 +1,307 @@
+even though the actual `ec2:DescribeInstances` call works (as long as it is
+against the allowed region).
+
+In order to acomodate installing OpenShift in these environments, a way is
+needed for the individual performing the installation to indicate that these
+permissions checks should be skipped.
+
+Additionally, a mechanism is needed to indicate to the cloud-credential-operator
+to force it into either `mint` or `passthrough` mode, so that it too can avoid
+attempting to validatate permissions.
+
+### Goals
+
+Enable successfull installation and operation of OpenShift in these AWS accounts
+where the results of permissions simulations cannot be relied upon.
+
+### Non-Goals
+
+Not looking to write complex policy introspection to implement what should
+already be performed by the AWS permissions simulation API.
+
+## Proposal
+
+### Installer
+Introduce an install-config.yaml field that can be populated by users to
+indicate that the installer should not concern itself with determining whether
+the credentials provided are sufficient for installation and to affect the
+in-cluster behavior of cloud-credential-operator.
+
+Extend the install-config type in the installer repo:
+```
+type InstallConfig struct {
+	// ForceCredentialsMode instructs the installer to not attempt to query
+	// the cloud permissions before attempting installation. It also passes
+	// down the desired credentials mode to the cloud-credential-operator
+	// so that it too does not attempt to query permissions.
+	// +optional
+	ForceCredentialsMode cloudcredop.CloudCredentialsMode `json:"forceCredentialsMode,omitempty"`
+}
+```
+
+The installer will then take the value defined in the `forceCredentialsMode`
+field to populate the config for the cloud-credential-operator so that CCO knows
+what mode it should be forced into.
+
+### cloud-credential-operator
+Formalize the constants in cloud-credential-operator repo to export:
+```
+type CloudCredentialsMode string
+
+const (
+	// MintMode is the annotation on the cluster's cloud credentials to
+	// indicate that CCO should be creating users for each
+	// CredentialsRequest.
+	MintMode CloudCredentialsMode = "mint"
+
+	// PassthroughMode is the annotation on the cluster's cloud credentials
+	// to indicate that CCO should just copy over the cluster's cloud
+	// credentials for each CredentialsRequest.
+	PassthroughMode CloudCredentialsMode = "passthrough"
+)
+```
+
+Introduce a config object to allow modifying the runtime behavior of CCO.
+
+```
+type CloudCredentialOperatorConfig struct {
+	Spec CloudCredentialOperatorConfigSpec `json:"spec"`
+}
+
+type CloudCredentialOperatorConfigSpec struct {
+	// ForceCredentialsMode will instruct CCO to skip any permissions
+	// checking and assume the designated mode when reconciling
+	// CredentialsRequests.
+	// +optional
+	ForceCredentialsMode CloudCredentialsMode `json:"forceCredentialsMode,omitempty"
+}
+```
```

Comment:
will/should it be required to follow the https://github.com/openshift/api/tree/master/operator/v1 conventions...

Reply:
Are you asking whether this new CRD will be committed to openshift/api (FWIW the CredentialsReqeust CRD is not in openshift/api today)?

Or are there some documented conventions for these objects (where?)? I can't immediately think of a reason why CCO would be special, so it should follow along with how the rest of our components are written.

Reply:
@abhinavdahiya yes, I can't think of a reason why we wouldn't follow existing config object conventions. I think it makes sense to move the CCO object type definitions into openshift/api in addition to the config object as a condition for completing this enhancement. @dgoodwin WDYT?

Reply:
Yeah moving our API over would be good.

I do read @abhinavdahiya 's comment as indicating there is some convention we should be following, it would be good to know what that is.

Reply:
I read it as simply following the patterns for the various config objects found in https://github.com/openshift/api/tree/master/operator/v1 (although I haven't found anywhere that formally documents these conventions)